lib/src/list.dart

```diff
@@ -66,64 +73,73 @@ class RealmList<T extends Object> extends collection.ListBase<T> {
+    if (items != null) {
+      _unmanaged.addAll(items);
+    }
   }
 
-  /// Gets a value indicating whether this collection is still valid to use.
-  ///
-  /// Indicates whether the [Realm] instance hasn't been closed,
-  /// if it represents a to-many relationship
-  /// and it's parent object hasn't been deleted.
-  bool get isValid => realmCore.listIsValid(this);
+  @override
+  int get length => _unmanaged.length;
+
+  @override
+  set length(int length) => _unmanaged.length = length;
+
+  @override
+  T operator [](int index) => _unmanaged[index] as T;
+
+  @override
+  void operator []=(int index, T value) => _unmanaged[index] = value;
+
+  @override
+  void clear() => _unmanaged.clear();
+
+  @override
+  bool get isValid => true;
 }
 
-// The query operations on lists only work for list of objects (core restriction),
-// so we add it as an extension method to allow the compiler to prevent misuse.
+// The query operations on lists, as well as the ability to subscribe for notifications,
+// only work for list of objects (core restriction), so we add these as an extension methods
+// to allow the compiler to prevent misuse.
 extension RealmListOfObject<T extends RealmObject> on RealmList<T> {
-  
-  /// Filters the list and returns a new [RealmResults] according to the provided query.
-  ///
-  /// Only works for lists of Realm objects.
-  ///
-  /// @param query The query used to filter the list
-  /// @param args Optional parameters for substitution in the query
+  /// Filters the list and returns a new [RealmResults] according to the provided [query] (with optional [arguments]).
   ///
-  /// @return The live result
+  /// Only works for lists of [RealmObject]s.
   ///
   /// The Realm Dart and Realm Flutter SDKs supports querying based on a language inspired by [NSPredicate](https://academy.realm.io/posts/nspredicate-cheatsheet/)
   /// and [Predicate Programming Guide.](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Predicates/AdditionalChapters/Introduction.html#//apple_ref/doc/uid/TP40001789)
-  ///
-  /// Only works for lists of objects.
-  RealmResults<T> query(String query, [List<Object> args = const []]) {
-    final handle = realmCore.queryList(this, query, args);
+  RealmResults<T> query(String query, [List<Object> arguments = const []]) {
+    final handle = realmCore.queryList(this, query, arguments);
     return RealmResultsInternal.create<T>(handle, realm);
   }
 
   /// Allows listening for changes when the contents of this collection changes.
   Stream<RealmListChanges<T>> get changes {
-    final controller = ListNotificationsController<T>(this);
+    final controller = ListNotificationsController<T>(managed);
     return controller.createStream();
   }
 }
 
 /// @nodoc
-extension RealmListInternal on RealmList {
-  RealmListHandle get handle => _handle;
+extension RealmListInternal<T extends Object> on RealmList<T> {
+  ManagedRealmList<T> get managed => this as ManagedRealmList<T>;
```

Comment:
I suggest we remove this getter. It allows very easy to forget and throw a type error somewhere.

Reply:
I see your point. I suggest making it private and renaming it to `_ensureManaged`, and have it throw a `RealmStateError` instead. The same has to happen a few places.

Reply:
Included in latest commit

Reply:
now called `asManaged()`